Configure Flutter CI

fkorotkov opened the pull request:
Cirrus CI recently added Flutter support by providing [Docker images with Flutter pre-installed](https://hub.docker.com/r/cirrusci/flutter/). It allows to avoid boilerplate of setting up an environment which makes it possible to start tests execution in a few hundred milliseconds from the moment a change is pushed to GitHub.

For this repository particularly I've seen CI times from 12 to 15 minutes for the longest part of the build (building apks). It's 2 to 3 times faster comparing to the current CI setup. 

Unfortunately `flutter_plugin_tools` can't run things in parallel and take advantage of up to 8 cores that Cirrus CI can provide.

Here is [a build for that particular change](https://cirrus-ci.com/build/5640120652791808):

![image](https://user-images.githubusercontent.com/989066/36113668-903d851c-0ffb-11e8-8811-4da945154ac6.png)

I was testing Flutter support on this repository and decided to share the results. Feel free to close the PR if not interested in a better CI. Didn't want to waste my work 😅

In case you will merge the PR. Don't forget to [install Cirrus CI application from Github Marketplace](https://github.com/apps/cirrus-ci) as described in [Quick Start guide](https://cirrus-ci.org/guide/quick-start/).

Hixie commented:
This is interesting. What version of Flutter does it test against?

cc @mravn-google

fkorotkov commented:
@Hixie it uses [`cirrusci/flutter:0.1.0` Docker container image](https://hub.docker.com/r/cirrusci/flutter/tags/) which has Flutter 0.1.0 pre-installed. There is 0.1.1 version available so I can bump it if you want.

BTW `cirrusci/flutter:0.1.0` is based on `cirrusci/android-sdk:26` image which has everything for Android development so it is possible to run an Android emulator using `cirrusci/flutter` images for integration tests.

fkorotkov commented:
cc @sethladd and @yjbanov since we have an ongoing email thread about Cirrus CI for Flutter projects.

Hixie commented:
We ship new versions of Flutter multiple times a week. I am worried about using a mechanism that is tied to a particular version. We care about testing against the latest version.

fkorotkov commented:
Fair point. Added `flutter channel master && flutter upgrade` which adds a little bit less than a minute to a build. Also removed `pub` cache to always test against the latest versions of libraries. See https://cirrus-ci.com/build/5708110253522944

yjbanov commented:
I'll be able to give this more thought after Feb 28. But I'm curious how well-trodden the Emulator/Simulator path is on Cirrus CI. This repository contains lots of Objective-C and Java code that need to be tested in either a simulated environment or real devices. Travis does not make it easy to test in emulators.

fkorotkov commented:
These images can [run an Android emulator](https://cirrus-ci.org/examples/#android). Unfortunately it won't have hardware acceleration since it's a Dockerized environment but, to my knowledge, no hosted CI has hardware accelerated emulators at the moment. Internally we have a design proposal on how to **securely** add such accelerated emulators but there are other priorities at the moment like adding Mac OS support.

Have a great vacation and let's revisit this PR once you are back 🙌

Hixie commented:
Ok I've landed this and turned on Cirrus CI! Let's see how it goes!

fkorotkov commented:
Thank you, @Hixie! Please let me know about any feedback and feel free to tag me on PRs for questions.
